fix inconsistent index naming with union/intersect #35847

iamlemec opened the pull request:
- [X] closes #35847
- [X] tests added / passed
- [X] passes `black pandas`
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

This takes care of some inconsistency in how names are handled by `Index` functions `union` and `intersection`, as discussed in #35847. I believe this covers all index types, either through the base class or in the subclass when necessary.

What I've implemented here actually uses the unanimous convention, wherein all input names must match to get assigned to the output. Originally, I was thinking consensus would be better (assign if there is only one non-None input name), but looking through the existing tests, it seems that unanimous was usually expected. I also had some worries about whether index names would become too "contagious" with consensus. Anyway, it's easy to change between the two if people have strong opinions on this.

iamlemec committed:
fix inconsistent index naming with union/intersect GH35847

jbrockmendel commented on pandas/core/indexes/base.py:
```diff
@@ -2654,12 +2658,16 @@ def _union(self, other, sort):
     def _wrap_setop_result(self, other, result):
         name = get_op_result_name(self, other)
-        return self._shallow_copy(result, name=name)
+        if isinstance(result, ABCIndexClass):
```
this check will include Index objects but exclude subclasses.  is that intentional?  if so, can you add a comment about why

iamlemec replied:
I'll admit I don't actually understand this whole ABC business 100%, but I'm getting:
```
isinstance(pd.MultiIndex, pd.Index) -> False
isinstance(pd.MultiIndex, ABCIndexClass) -> True
issubclass(pd.MultiIndex, pd.Index) -> True
```
Should I be using `issubclass` instead? Wasn't trying to do anything too fancy here.

jbrockmendel replied:
yah its not intuitive.  basically `isinstance(obj, ABCIndexClass)` is equivalent to `type(obj) is Index`, so what you probably want is `isinstance(obj, Index)`, which behaves like normal

iamlemec replied:
Hmm, but wouldn't we get `isinstance(multiindex, ABCIndexClass) == True` and `type(multiindex) != pd.Index` here? Either way, agreed that `isinstance` will get us what we want.

jbrockmendel replied:
my bad, i confused ABCIndexClass with ABCIndex, never mind.

jbrockmendel commented on pandas/core/indexes/base.py:
```diff
@@ -2654,12 +2658,16 @@ def _union(self, other, sort):
         name = get_op_result_name(self, other)
-        return self._shallow_copy(result, name=name)
+        if isinstance(result, ABCIndexClass):
+            if result.name != name:
```
if you want to be really thorough, could handle cases with names like `np.nan` which is not equal to itself.  (but since we're not good about that anyway, OK to leave out-of-scope)

iamlemec replied:
Yeah, I saw a note about that in `get_op_result_name` too. I can look into it after this.

jbrockmendel commented on pandas/core/indexes/base.py:
```diff
@@ -5926,3 +5931,22 @@ def _maybe_asobject(dtype, klass, data, copy: bool, name: Label, **kwargs):
     return klass(data, dtype=dtype, copy=copy, name=name, **kwargs)
+
+
+def get_unanimous_names(*indexes):
```
do we ever get more than two here?

Would `indexes` instead of `*indexes` work>?

iamlemec replied:
yeah, we use this in `union_indexes` which operates on a list of indices. but most uses are explicitly two, so I figured this would make the typical syntax easier.

jreback replied:
can you type the input and output

jbrockmendel commented on pandas/core/indexes/datetimelike.py:
```diff
@@ -670,33 +670,29 @@ def intersection(self, other, sort=False):
+            return self.copy()._get_reconciled_name_object(other)
         if len(other) == 0:
-            return other.copy()
+            return other.copy()._get_reconciled_name_object(self)
```
in Index._union we dont copy, probably should

iamlemec replied:
The danger here is that someone accidentally modified the original version? Want me to add that to `Index._union` or save for later?

jbrockmendel replied:
lets go ahead and make it consistent now, yah

jbrockmendel commented on pandas/core/indexes/multi.py:
```diff
@@ -3415,7 +3420,7 @@ def union(self, other, sort=None):
 
         if len(other) == 0 or self.equals(other):
-            return self
+            return self._shallow_copy(names=result_names)
```
self.rename?

iamlemec replied:
Yup, makes sense. Certainly more legible, and seems like equivalent.

jbrockmendel commented on pandas/core/indexes/numeric.py:
```diff
@@ -180,7 +180,8 @@ def _union(self, other, sort):
             second = other.astype("float")
-            return first._union(second, sort)
+            result = first._union(second, sort)
+            return Float64Index(result)
```
this should already be a Float64Index.   are there cases where it isnt?

iamlemec replied:
Ah, you're right. At some point I was allowing `_union` to return just an array, so I ended up needing that. But now there's a `_shallow_copy` at return, so it's redundant. Will revert.

jbrockmendel commented on pandas/tests/indexes/multi/test_setops.py:
```diff
@@ -242,10 +242,10 @@ def test_union(idx, sort):
 
     the_union = idx.union(idx[:0], sort=sort)
-    assert the_union is idx
+    assert tm.equalContents(the_union, idx)
```
can you use tm.assert_index_equal

iamlemec replied:
done

jreback commented on pandas/core/indexes/api.py:
```diff
@@ -58,6 +59,7 @@
     "get_objs_combined_axis",
     "union_indexes",
     "get_consensus_names",
+    "get_unanimous_names",
```
can we totally remove get_consenus_names? (alt could cal it `get_result_name` and use these routines under the hood)

iamlemec replied:
Only other place `get_consensus_names` is used is in `concat` for the nested MultiIndex case. In fact, looking at master, this ends up dropping level names even when they all agree since it operates jointly on all levels, rather than going level-by-level like `get_unanimous_names`. Swapping in `get_unanimous_names` here lets us totally drop `get_consensus_names` and seems to work very well, all tests passed.

jreback commented on doc/source/whatsnew/v1.2.0.rst:
```diff
@@ -293,6 +293,7 @@ Indexing
 
 - Bug in :meth:`PeriodIndex.get_loc` incorrectly raising ``ValueError`` on non-datelike strings instead of ``KeyError``, causing similar errors in :meth:`Series.__geitem__`, :meth:`Series.__contains__`, and :meth:`Series.loc.__getitem__` (:issue:`34240`)
 - Bug in :meth:`Index.sort_values` where, when empty values were passed, the method would break by trying to compare missing values instead of pushing them to the end of the sort order. (:issue:`35584`)
+- Harmonize resulting index names from :meth:`Index.union` and :meth:`Index.intersection` across various index types (:issue:`35847`)
```
if there are significant changes that a user would want to know can you add that detail

jreback replied:
also migth warrant a sub-section if this is something easier to see via code-examples

iamlemec replied:
yup, added in a bit about name preservation

jreback replied:
right add the issue number above and remove this note

jreback commented on pandas/tests/indexes/test_common.py:
```diff
@@ -124,6 +124,65 @@ def test_corner_union(self, index, fname, sname, expected_name):
         expected = index.drop(index).set_names(expected_name)
         tm.assert_index_equal(union, expected)
 
+        # test copy.union(subset) - need sort for unicode and string
```
can you make this another test

iamlemec replied:
done

jreback reviewed (changes requested):
looks really good @iamlemec 

only small concern is that we are likely not fully validating whether we are returning a copy for the 0 len cases or where indexes are equal.

but certainly can do as a followup.

iamlemec commented:
Ok, added in a code example to whatsnew showing the new/consistent naming behavior when aggregating. Also split off those new tests.

Regarding copying in the various empty/equality cases, copy here means a deep copy including values, right? Seems like some of the tests are using "is" which breaks when you add in copying. I can look into that and the nan issues in a subsequent PR if that works.

jreback commented on doc/source/whatsnew/v1.2.0.rst:
```diff
@@ -109,6 +109,25 @@ Beginning with this version, the default is now to use the more accurate parser
+Index/column name preservation when aggregating
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+When aggregating using :meth:`concat` or the :class:`DataFrame` constructor, Pandas
```
can you add the issue reference here.

Also I would add the same text here in the docs, maybe in the reshaping section, in a note

iamlemec replied:
yep, i ended up adding it as a note in the "merge, join, concatenate, and compare" section.

jreback reviewed (changes requested):
doc comment, ping on green.

jbrockmendel commented on pandas/core/indexes/datetimes.py:
```diff
@@ -383,6 +383,10 @@ def union_many(self, others):
+
+        res_name = get_unanimous_names(self, *others)[0]
+        if this.name != res_name:
+            return this._shallow_copy(name=res_name)
```
rename instead of shallow_copy?

iamlemec replied:
yup, indeed. i actually went through and made the same change in `base.py` and `api.py`. seems like it works fine all around.

iamlemec commented:
@jreback fixed the docs indentation issue, that's green now. i'm having trouble figuring out went wrong with some of the other tests. is it some kind of internal CI error?

jbrockmendel commented:
CI failures were unrelated, should be fixed if you merge master

jbrockmendel commented on pandas/core/indexes/base.py:
```diff
@@ -5935,3 +5941,22 @@ def _maybe_asobject(dtype, klass, data, copy: bool, name: Label, **kwargs):
     return klass(data, dtype=dtype, copy=copy, name=name, **kwargs)
+
+
+def get_unanimous_names(*indexes: Type[Index]) -> Tuple[Any, ...]:
```
`Type[Index]` seems really weird here.  Shouldnt it be `Tuple[Index]`?  And i think the return type should use Label instead of Any

iamlemec replied:
Ah, I was wondering if there was a `Label` notion! And now I see there's a `_typing.py`, will keep that handy in the future.

Yeah, I don't know why that extraneous ended up `Type` there. But actually with an *args situation, you just give the type of the individual elements, so I think it should just be `Index`.

jbrockmendel commented on pandas/core/indexes/datetimelike.py:
```diff
@@ -719,33 +719,29 @@ def intersection(self, other, sort=False):
 
         if len(self) == 0:
-            return self.copy()
+            return self.copy()._get_reconciled_name_object(other)
```
in the base class you got rid of the `_get_reconciled_name_object` usage.  why going the other direction here?

iamlemec replied:
Still keeping this around for the rare case where you want to return the same values but with a possible name change. In base.py it only shows up in the one `intersection` return clause (and not in `union` because of the `union`/`_union` split). It is interesting though that the version of `intersection` in datetimelike.py has two more quick return clauses than the base class.

iamlemec commented:
I think we're mostly good here, but the travis-ci build for ARM just kinda bailed without comment.

jbrockmendel commented on pandas/core/reshape/concat.py:
```diff
@@ -23,10 +23,10 @@
     get_objs_combined_axis,
 )
 import pandas.core.indexes.base as ibase
+from pandas.core.indexes.base import get_unanimous_names
```
get_unanimous_names is in indexes.api, should get it from there

jbrockmendel commented on pandas/tests/indexes/datetimes/test_setops.py:
```diff
@@ -471,7 +471,7 @@ def test_intersection_bug(self):
-        idx = pd.DatetimeIndex(values, name="a")
+        idx = pd.DatetimeIndex(values)
         res = idx.intersection(values)
         tm.assert_index_equal(res, idx)
```
could keep the name in `idx` and test for `tm.assert_index_equal(res, idx.rename(None))`?

jbrockmendel commented on pandas/tests/indexes/multi/test_join.py:
```diff
@@ -46,7 +46,7 @@ def test_join_level_corner_case(idx):
 def test_join_self(idx, join_type):
     joined = idx.join(idx, how=join_type)
-    assert idx is joined
+    assert tm.equalContents(joined, idx)
```
does assert_index_equal not work here?

jbrockmendel commented on pandas/tests/indexes/multi/test_setops.py:
```diff
@@ -278,7 +278,7 @@ def test_intersection(idx, sort):
     # corner case, pass self
     the_int = idx.intersection(idx, sort=sort)
-    assert the_int is idx
+    assert tm.equalContents(the_int, idx)
```
assert_index_equal?

iamlemec commented:
Thanks @jbrockmendel! Pushed all your suggested changes.

jbrockmendel commented:
@iamlemec can you merge master; IIRC this was pretty close to ready

iamlemec commented:
@jbrockmendel sure thing, just pushed a rebase onto master

jreback commented:
thanks @iamlemec very nice!

iamlemec commented:
thanks for all the help @jreback and @jbrockmendel!
